### DISCUSSION TITLE
400 Bad Request - Unable to download the mesh agent from the RMM

### DISCUSSION THREAD

**Discussion by sport80** [amidaware/tacticalrmm#1540] [Q&A]
**Server Info:**

OS: Debian 5.10.179-1 (2023-05-12) x86_64
Browser: NA
RMM Version: v0.15.12

**Installation Method:**
[ x ] Standard
 Docker

**Agent Info :**
Agent version : Agent v2.4.9
Agent OS: Windows Server 2016
Describe the bug
The installation of the agent fail with the following error : "Unable to download the mesh agent from the RMM"

**To Reproduce**

Download Dynamically generated exe or Powershell or Manual install script
Go to on the computer
Launch the install script
Installation fail when the RMM is downloaded

**Expected behavior**
Agent should properly setup and computer display in the RMM

**Description:**
Hi, i've installed a new VM to try and deploy your TacticalRMM solution. I've modified the installation script commenting the parts that generate a let's encrypt certificate because we already have our internal certificate. The VM have the CA correctly installed and the wildcard certificate mapped in the script, so, the installation was ending correctly without errors.

At the first login I've tried to install the first agent, so, i've generated the agent and uploaded to the windows server, but the installation fails with the error described before.

Using the manual installation with debug mode this is the results of error:

![image](https://github.com/amidaware/tacticalrmm/assets/7521521/0ff7e9f1-f986-4954-b185-7744730fe2c9)

I've already read the previous post atbout this error, but i've not found a solution. The installation appears to be correct

![image](https://github.com/amidaware/tacticalrmm/assets/7521521/f8970ca2-245b-45d4-b083-341a350b8be1)

Thank you for your support

Regards

**Comment by dinger1986** [comment]
did you follow the guide for using your own SSL cert? if not that will have caused issues

**Reply by sport80** [reply]
Hi, thank you for your reply

This one, right? https://docs.tacticalrmm.com/functions/settings_override/#using-your-own-wildcard-ssl-cert

I've already edited the install script so it use the custom SSL cert

```
print_green 'Getting wildcard cert'

#sudo certbot certonly --manual -d *.${rootdomain} --agree-tos --no-bootstrap --preferred-challenges dns -m ${letsemail} --no-eff-email
#while [[ $? -ne 0 ]]; do
#  sudo certbot certonly --manual -d *.${rootdomain} --agree-tos --no-bootstrap --preferred-challenges dns -m ${letsemail} --no-eff-email
#done

#CERT_PRIV_KEY=/etc/letsencrypt/live/${rootdomain}/privkey.pem
#CERT_PUB_KEY=/etc/letsencrypt/live/${rootdomain}/fullchain.pem
CERT_PRIV_KEY=/etc/ssl/domain.local/domain.local.key
CERT_PUB_KEY=/etc/ssl/domain.local/domain.local.crt

#sudo chown ${USER}:${USER} -R /etc/letsencrypt

print_green 'Installing Nginx'

wget -qO - https://nginx.org/packages/keys/nginx_signing.key | sudo apt-key add -
```

It's a fresh installation, not an upgrade. I also need to edit the file /rmm/api/tacticalrmm/tacticalrmm/local_settings.py to set the CERT key and file?

Thanks

**Reply by sport80** [reply]
I've edited the file  /rmm/api/tacticalrmm/tacticalrmm/local_settings.py and i've found that the two lines CERT_FILE = and KEY_FILE = are missing.

I've added the two lines and specified the full path to the certs but nothing changed. The agent return errors on download the mesh agent.

Thanks

**Reply by sport80** [reply]
* nats reloaded using /rmm/api/env/bin/python /rmm/api/tacticalrmm/manage.py reload_nats
* server reboot
* the cert files are readable by the tactical user

**Reply by dinger1986** [reply]
did you run `update.sh --force`

**Reply by dinger1986** [reply]
does rmm.yourdomain.com work? what about mesh.yourdomain.com?

Best option would be to install using letsencrypt wildcard and then replace it once its done. You can have multiple wildcards for a domain so its fine to do.

Anything else is spending alot of time troubleshooting something you have done differently, I dont know if there would be an issue not using pem files as thats all that I have ever used

**Reply by sport80** [reply]
> does rmm.yourdomain.com work? what about mesh.yourdomain.com?

Yes, it works correctly. 

> Best option would be to install using letsencrypt wildcard and then replace it once its done. You can have multiple wildcards for a domain so its fine to do.
> 
> Anything else is spending alot of time troubleshooting something you have done differently, I dont know if there would be an issue not using pem files as thats all that I have ever used

Sorry, but this is a non-sense, IMHO. So, If i need only to deploy a local test VM or if i don't have a public domain i can't install tacticalrmm? This is a significant limitation...

**Reply by dinger1986** [reply]
NATs (which is what is used for agents to talk to the server) needs genuine SSL certs, doesn’t mean you can’t have it local, just use a real domain and setup the 1 text record to verify to get the wildcard cert, then once it’s working follow the docs and swap it across to your own cert.

It’s not about whether you can install it or not (or if it will work), it’s how much time (for free) does someone need to support you through that process.

**Reply by sport80** [reply]
Your project is open, it's your choice. The problem that I have could also be experienced by other users and this would help the project to grow and to improve. It's the first time in more than 20 years of management, forums and support on open source projects that someone accuses me of the **free time** made available. Thank you.

**Reply by dinger1986** [reply]
I also explained a way to make it work and tried to explain why I was suggesting doing it the way I said. No accusation was intended!

**Comment by NiceGuyIT** [comment]
> The problem that I have could also be experienced by other users and this would help the project to grow and to improve.

Opinionated projects grow fast because they provide a limited path to do things which translates to an easy support path. 

> Sorry, but this is a non-sense, IMHO. So, If i need only to deploy a local test VM or if i don't have a public domain i can't install tacticalrmm?

Here's the link to NATS docs for using [Self Signed Certificates for Testing][] and [Enabling TLS][] on the server. The client will need to verify the TLS chain in order for it to connect. If you know TLS, then you know you can create a TLS CA, create TLS certs using that CA, add the CA root to all computers (servers and workstations) involved. You are providing your own PKI and it will work.

There are docs explaining the [MeshCentral Integration][] with Tactical RMM. In case you don't know, NATS is for TRMM, not MeshCentral. [MeshCentral][] is a separate project and has different TLS configuration.

> support on open source projects that someone accuses me of the **free time** made available

Considering you opened a discussion on Tactical RMM to ask for help for an unsupported install, @dinger1986's response is completely reasonable. You have no idea how much time is spent trying to help people with **their environment**, not with Tactical RMM. It ends up being a training session on how networking works, or TLS works, or reverse proxies. Some people don't want to learn, they just want it to work. Some people don't want to pay for commercial support because "it's for my home lab", while others will gladly pay because they use it for their business. Your viewpoint seems to be that of "Why can't I install the software my way and make it work?" Try viewing it from the other viewpoint: you are one of hundreds, if not thousands, of people asking for _free support_ to make the software fit your environment. When it's suggested to use the well worn path of support by using a real domain, you become aggressive because "This is a significant limitation...". No. It's not. It's meant to provide a happy path for installation and an easy path for support (to save their sanity!).

[Self Signed Certificates for Testing]: https://docs.nats.io/running-a-nats-service/configuration/securing_nats/tls#self-signed-certificates-for-testing
[Enabling TLS]: https://docs.nats.io/running-a-nats-service/configuration/securing_nats/tls
[MeshCentral Integration]: https://docs.tacticalrmm.com/mesh_integration/
[MeshCentral]: https://github.com/Ylianst/MeshCentral/

**Reply by sport80** [reply]
I have asked because following your FAQ this is a supported installation method, not a customized or what else installation. https://docs.tacticalrmm.com/functions/settings_override/#using-your-own-wildcard-ssl-cert

> Your viewpoint seems to be that of "Why can't I install the software my way and make it work?"

No, this not my viewpoint. I also have supported more open source projects in past and so i know what does it mean.

> When it's suggested to use the well worn path of support by using a real domain, you become aggressive because "This is a significant limitation...".

Do you really think that this is an aggressive response? No comment...  You don't know the difference between a consideration and an offense.

**Comment by silversword411** [comment]
> If i need only to deploy a local test VM or if i don't have a public domain i can't install tacticalrmm? This is a significant limitation...

If you want a local test install with broken or self signed certs, you can do a dev install and manage all certs on all machines end-to-end, and you'll need to work around all the problems of using untrusted encryption everywhere. 

If you want to have a supported installation that just works, you have to follow all the requirements of TRMM
https://docs.tacticalrmm.com/install_server/#minimum-requirements
https://docs.tacticalrmm.com/install_server/#network-requirements

Yes, that includes a paid for public domain name. Cloudflare can get you one for $10/year.

Does that mean you have to expose your TRMM server to the internet? No
https://docs.tacticalrmm.com/faq/#can-i-run-tactical-rmm-locally-behind-nat-without-exposing-my-rmm-server-to-the-internet

Is it also correct that if you don't have minimum requirements you can't install TRMM easily and just have it work, you're trying to do an unsupported install.

We have a pinned Discord item about TRMM and the types of installs which describes the differences nicely. 

The difference between ⁠supported and ⁠unsupported (and wizards):
Supported: We troubleshoot the application. This helps improve the application and documentation in general. Everybody wins.
Unsupported: We troubleshoot your environment. This helps only the individual because everyone has a unicorn setup. One person wins.
Wizards: They don't ask for support because they know their environment, know how to troubleshoot systems and applications, and can make TRMM bend to their will!

**Comment by sport80** [comment]
> If you want a local test install with broken or self signed certs, you can do a dev install and manage all certs on all machines end-to-end, and you'll need to work around all the problems of using untrusted encryption everywhere.

Thank you. For testing I've deployed the certificate through an active directory domain controller to the clients. At this moment it works using https, tls, ssl, mssql session and more over.

I'll try investigantig, but i don't know the architecture of tacticalrmm so, for me, reading the log it try to connect using an https session that calling it manually is solved correctly, but using the agent not.

Maybe i can find more information using WireShark...

**Reply by silversword411** [reply]
Lots of resources here
https://docs.tacticalrmm.com/howitallworks/

**Reply by andjeskiii** [reply]
Hello Sport80, 

I have the same problem, did you succeed ? ?

**Reply by sport80** [reply]
Yes, I've solved the problem, but I've abandoned the deploy because the Linux client it's only via payment option. As a fairly animated discussion emerged and with considerations that I didn't like, I decided to use another product.

**Reply by andjeskiii** [reply]
how did you solve it? can you please tell me. I'm doing a project for school.
